buildSrc/src/main/java/org/elasticsearch/gradle/VersionCollection.java

```diff
@@ -0,0 +1,331 @@
+            }
+        }
+
+        return unmodifiableList(
+            unreleased.stream()
+                .sorted()
+                .distinct()
+                .collect(Collectors.toList())
+        );
+    }
+
+    private Version getLatestVersionByKey(Map<Integer, List<Version>> groupByMajor, int key) {
+        return groupByMajor.getOrDefault(key, emptyList()).stream()
+            .max(Version::compareTo)
+            .orElseThrow(() -> new IllegalStateException("Unexpected number of versions in collection"));
+    }
+
+    private Map<Integer, List<Version>> getReleasedMajorGroupedByMinor() {
+        List<Version> currentMajorVersions = groupByMajor.get(currentVersion.getMajor());
+        List<Version> previousMajorVersions = groupByMajor.get(currentVersion.getMajor() - 1);
+
+        final Map<Integer, List<Version>> groupByMinor;
+        if (currentMajorVersions.size() == 1) {
+            // Current is an unreleased major: x.0.0 so we have to look for other unreleased versions in the previous major
+            groupByMinor = previousMajorVersions.stream()
+                .collect(Collectors.groupingBy(Version::getMinor, Collectors.toList()));
+        } else {
+            groupByMinor = currentMajorVersions.stream()
+                .collect(Collectors.groupingBy(Version::getMinor, Collectors.toList()));
+        }
+        return groupByMinor;
+    }
+
+    public void compareToAuthoritative(List<Version> authoritativeReleasedVersions) {
+        Set<Version> notReallyReleased = new HashSet<>(getReleased());
+        notReallyReleased.removeAll(authoritativeReleasedVersions);
+        if (notReallyReleased.isEmpty() == false) {
+            throw new IllegalStateException(
+                "out-of-date released versions" +
+                    "\nFollowing versions are not really released, but the build thinks they are: " + notReallyReleased
+            );
+        }
+
+        Set<Version> incorrectlyConsideredUnreleased = new HashSet<>(authoritativeReleasedVersions);
+        incorrectlyConsideredUnreleased.retainAll(getUnreleased());
+        if (incorrectlyConsideredUnreleased.isEmpty() == false) {
+            throw new IllegalStateException(
+                "out-of-date released versions" +
+                    "\nBuild considers versions unreleased, " +
+                    "but they are released according to an authoritative source: " + incorrectlyConsideredUnreleased +
+                    "\nThe next versions probably needs to be added to Version.java (CURRENT doesn't count)."
+            );
+        }
+    }
+
+    private List<Version> getReleased() {
+        List<Version> unreleased = getUnreleased();
+        return groupByMajor.values().stream()
+            .flatMap(Collection::stream)
+            .filter(each -> unreleased.contains(each) == false)
+            .sorted()
+            .collect(Collectors.toList());
+    }
+
+    public List<Version> getIndexCompatible() {
```

Comment:
These all seem like others that could be precomputed in the ctor (index compatible, released, wire compatible, etc)